### PR TITLE
Refactor types: extract API for `ttc` and `tt` ##refactor

### PR DESCRIPTION
**Pull request by officialcjunior**
**Your checklist for this pull request**
- [x] I've read the [guidelines for contributing](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md) to this repository
- [x] I made sure to follow the project's [coding style](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md#code-style)
- [x] I've added tests that prove my fix is effective or that my feature works (if possible)
- [ ] I've updated the documentation and the [rizin book](https://github.com/rizinorg/book) with the relevant information (if needed)

**Detailed description**
This PR refactors a handful of places in `cmd_type.c` to use the API.
I've added two new RZ_APIs for the commands `tt` and `ttc`.
And they are:
```
RZ_API void rz_core_list_typename_alias_c(RzCore *core, const char *typedef_name);
RZ_API void rz_core_list_loaded_typedefs(RzCore *core, RzOutputMode mode);
```
Also added tests for `ttj`
**Test plan**
Firstly, we need to,
- [x] Name the APIs correctly and put them in the right place.

I'm not sure whether I've made `tc` working fine. I am currently looking into that.
But anyway, all of them have written unit tests, so if the CI succeeds, we can say that they're working fine.

**Closing issues**

None.

### CODE REVIEW

File: librz/core/cmd.c [hunk 1]
```diff
@@ -380,7 +380,8 @@ static bool lastcmd_repeat(RzCore *core, int next) {
 	case '$':
 		if (!strncmp(core->lastcmd, "pd", 2)) {
 			if (core->lastcmd[2] == ' ') {
-				rz_core_cmdf(core, "so %s", core->lastcmd + 3);
+				ut64 addr = rz_num_math(core->num, core->lastcmd + 3);
```
Comment (ret2libc) [inline librz/core/cmd.c:383]: this change is ok, though AFAIK the whole logic here is broken.

File: librz/core/cmd_type.c [hunk 2]
```diff
@@ -1010,15 +1076,15 @@ RZ_IPI int rz_cmd_type(void *data, const char *input) {
 			if (name && type) {
 				name++; // skip the '.'
 				if (rz_str_startswith(type, "struct")) {
-					rz_core_cmdf(core, "tsc %s", name);
+					print_struct_union_in_c_format(TDB, stdifstruct, rz_str_trim_head_ro(input + 2), true);
```
Comment (ret2libc) [inline librz/core/cmd_type.c]: shouldn't you just use `name`?
```suggestion
					print_struct_union_in_c_format(TDB, stdifstruct, name, true);
```

File: librz/core/cmd_type.c [hunk 3]
```diff
@@ -1010,15 +1076,15 @@ RZ_IPI int rz_cmd_type(void *data, const char *input) {
...
 				} else if (rz_str_startswith(type, "union")) {
-					rz_core_cmdf(core, "tuc %s", name);
+					print_struct_union_in_c_format(TDB, stdifunion, rz_str_trim_head_ro(input + 2), true);
```
Comment (ret2libc) [inline librz/core/cmd_type.c]: same

File: librz/core/cmd_type.c [hunk 4]
```diff
@@ -1010,15 +1076,15 @@ RZ_IPI int rz_cmd_type(void *data, const char *input) {
...
 				} else if (rz_str_startswith(type, "enum")) {
-					rz_core_cmdf(core, "tec %s", name);
+					print_enum_in_c_format(TDB, rz_str_trim_head_ro(input + 2), true);
```
Comment (ret2libc) [inline librz/core/cmd_type.c]: same

File: librz/core/cmd_type.c [hunk 5]
```diff
@@ -1010,15 +1076,15 @@ RZ_IPI int rz_cmd_type(void *data, const char *input) {
...
 				} else if (rz_str_startswith(type, "typedef")) {
-					rz_core_cmdf(core, "ttc %s", name);
+					rz_core_list_typename_alias_c(core, input, TDB);
```
Comment (ret2libc) [inline librz/core/cmd_type.c]: same

File: librz/core/cmd_type.c (hunk 2 above)
Comment (officialcjunior) [reply librz/core/cmd_type.c]: Ah, sorry, yes.

**Comment by XVilka** [comment]
Don't we have tests for those? Could you please add a couple? Will be helpful for the future RzTypes refactor as well.

File: librz/core/cmd_type.c [hunk 6]
```diff
@@ -1010,15 +1076,15 @@ RZ_IPI int rz_cmd_type(void *data, const char *input) {
...
+					print_struct_union_in_c_format(TDB, stdifunion, name, true);
 				} else if (rz_str_startswith(type, "enum")) {
-					rz_core_cmdf(core, "tec %s", name);
+					print_enum_in_c_format(TDB, name, true);
 				} else if (rz_str_startswith(type, "typedef")) {
-					rz_core_cmdf(core, "ttc %s", name);
+					rz_core_list_typename_alias_c(core, input, TDB);
 				} else if (rz_str_startswith(type, "func")) {
-					rz_core_cmdf(core, "tfc %s", name);
+					printFunctionTypeC(core, input + 3);
```
Comment (ret2libc) [inline librz/core/cmd_type.c]: name here as well?

**Comment by officialcjunior** [comment]
> Don't we have tests for those? Could you please add a couple? Will be helpful for the future RzTypes refactor as well.

Added tests for `tfc` and `tsc`.
Do you know where I could find some enums and typenames/typedefs in the testbins, so that I could write tests for `ttc` and `tuc`?

File: librz/core/cmd_type.c (hunk 5 above)
Comment (officialcjunior) [reply librz/core/cmd_type.c]: Missed this. Will fix in a later commit.

**Comment by officialcjunior** [comment]
> Don't we have tests for those? Could you please add a couple? Will be helpful for the future RzTypes refactor as well.
Which commands were you referring to?

Actually, I just realized that `tsc`, `tfc`, `tuc` and `tec` already had tests in db/cmd/types, which are fr the ones I just added.
So, should I revert these changes or keep it?

**Comment by ret2libc** [comment]
Please mark it as "ready to review" when we can have a look ;)

**Review by XVilka — APPROVED** [review]

File: librz/core/cmd_type.c [hunk 7]
```diff
@@ -1627,78 +1693,16 @@ RZ_IPI int rz_cmd_type(void *data, const char *input) {
...
-			if (input[1] == 'j') {
-				pj_end(pj);
-			}
-			if (pj) {
-				rz_cons_printf("%s\n", pj_string(pj));
-				pj_free(pj);
-			}
-			free(name);
-			ls_free(l);
+			rz_core_list_loaded_typedefs(core, input, TDB);
```
Comment (ret2libc) [inline librz/core/cmd_type.c]: this API is not good. We should avoid creating API that requires a  general "input" and keep parsing the input. The input should be parsed by the command handler and converted to the more appropriate type, then a more generic API should be called. For example, use RzOutputMode to indicate whether JSON or not.

**Review by ret2libc — CHANGES_REQUESTED** [review]

File: librz/core/cmd_type.c (hunk 7 above)
Comment (XVilka) [reply librz/core/cmd_type.c]: Right, missed that.

Comment (officialcjunior) [reply librz/core/cmd_type.c]: Right. 
I have looked at another particular API and modified it accordingly to use `RzOutputMode` for JSON and all. 
Please see if that's what you had meant.
Thanks for the catching that, by the way. Learned new things while doing that!

File: librz/core/cmd_type.c [hunk 8]
```diff
@@ -729,6 +725,76 @@ static void set_offset_hint(RzCore *core, RzAnalysisOp *op, const char *type, ut
...
+					const char *res = sdb_const_get(TDB, q, 0);
+					pj_ks(pj, name, res);
+				}
+			}
+		}
+	}
+	if (mode == RZ_OUTPUT_MODE_JSON) {
+		pj_end(pj);
+	}
+	if (pj) {
```
Comment (ret2libc) [inline librz/core/cmd_type.c]: just put the code inside this if in the body of the if above.

File: librz/core/cmd_type.c [hunk 9]
```diff
@@ -729,6 +725,76 @@ static void set_offset_hint(RzCore *core, RzAnalysisOp *op, const char *type, ut
...
+	char *name = NULL;
+	SdbKv *kv;
+	SdbListIter *iter;
+	SdbList *l = sdb_foreach_list(TDB, true);
+	ls_foreach (l, iter, kv) {
+		if (!strcmp(sdbkv_value(kv), "typedef")) {
+			if (!name || strcmp(sdbkv_value(kv), name)) {
+				free(name);
+				name = strdup(sdbkv_key(kv));
+				if (!input[1]) {
```
Comment (ret2libc) [inline librz/core/cmd_type.c]: this check was just used to see which output mode had to be used. You can just check `mode` instead. Also, by removing this, you can also remove the `input` argument. By doing so, you make the whole function cleaner and easier to use and understand.

File: librz/core/cmd_type.c [hunk 10]
```diff
@@ -729,6 +725,76 @@ static void set_offset_hint(RzCore *core, RzAnalysisOp *op, const char *type, ut
 	}
 }
 
+RZ_API void rz_core_list_loaded_typedefs(RzCore *core, const char *input, Sdb *TDB, RzOutputMode mode) {
```
Comment (ret2libc) [inline librz/core/cmd_type.c]: TDB is not needed.
If you see where this is called, TDB is just `	Sdb *TDB = core->analysis->sdb_types;`, so just take it from the `core` argument, no need for an additional arg.

File: librz/core/cmd_type.c [hunk 11]
```diff
@@ -729,6 +725,76 @@ static void set_offset_hint(RzCore *core, RzAnalysisOp *op, const char *type, ut
...
+	}
+	if (pj) {
+		rz_cons_printf("%s\n", pj_string(pj));
+		pj_free(pj);
+	}
+	free(name);
+	ls_free(l);
+}
+
+RZ_API void rz_core_list_typename_alias_c(RzCore *core, const char *input, Sdb *TDB) {
```
Comment (ret2libc) [inline librz/core/cmd_type.c]: no need for `TDB`, as above.

Comment (ret2libc) [inline librz/core/cmd_type.c]: Instead of `input` change name to something more meaningful, like `typedef_name`

**Review by ret2libc — CHANGES_REQUESTED** [review]

File: librz/core/cmd_type.c [hunk 12]
```diff
@@ -1627,81 +1696,26 @@ RZ_IPI int rz_cmd_type(void *data, const char *input) {
...
+			rz_core_list_typename_alias_c(core, input);
 			break;
 		}
-		if (input[1] == '?') {
+		if (input[1] == '?') { // "tt?"
 			rz_core_cmd_help(core, help_msg_tt);
 			break;
 		}
+		if (input[1] == ' ') { // "tt"
+			mode = RZ_OUTPUT_MODE_STANDARD;
```
Comment (ret2libc) [inline librz/core/cmd_type.c]: just pass it to `rz_core_list_loaded_typedefs`.

File: librz/core/cmd_type.c [hunk 13]
```diff
@@ -1627,81 +1696,26 @@ RZ_IPI int rz_cmd_type(void *data, const char *input) {
...
+			rz_core_list_loaded_typedefs(core, mode);
```
Comment (ret2libc) [inline librz/core/cmd_type.c]:
```suggestion
			rz_core_list_loaded_typedefs(core, RZ_OUTPUT_MODE_STANDARD);
```

**Review by ret2libc — APPROVED** [review]

**Review by XVilka — CHANGES_REQUESTED** [review]
One failing test:
```
[XX] db/cmd/types tt
RZ_NOPLUGINS=1 rizin -escr.utf8=0 -escr.color=0 -escr.interactive=0 -N -Qc 'td typedef char FILE_NAME;
tt FILE_NAME
"td typedef bool FLAG;"
tt FLAG
' -
-- stdout
```

**Comment by XVilka** [comment]
It's because `tt` commands expects an argument, and refactoring removed that. See the test - `tt FLAG` should return `bool`, `tt FILE_NAME` should return `char`.

**Review by ret2libc — APPROVED** [review]

**Review by XVilka — APPROVED** [review]

**Pull request merged**